File: SearchEngine.cpp
#include "SearchEngine.hpp"
#include <algorithm>
#include <iostream>

using namespace std;

// public methods
vector <Book> SearchEngine::searchByTitle(const vector<Book>& books, const std::string& query) const
{
    vector<Book> results;
    for (const auto& book : books) {
        if (book.getTitle().find(query) != string::npos) {
            results.push_back(book);
        }
    }
    return results;
}

vector <Book> SearchEngine::searchByAuthor(const vector<Book>& books, const std::string& query) const
{
    vector<Book> results;
    for (const auto& book : books) {
        if (book.getAuthor().find(query) != string::npos) {
            results.push_back(book);
        }
    }
    return results;
}

File: CourceWork2Java/src/SearchEngine.java
import java.util.ArrayList;
import java.util.List;

public class SearchEngine {

    // Constructors
    public SearchEngine() {
        // System.out.println("SearchEngine: Constructor called.");
    }

    // Public methods
    public List<Book> searchByTitle(List<Book> books, String query) {
        List<Book> results = new ArrayList<>();
        if (query == null || query.isEmpty()) return results; // Empty request

        for (Book book : books) {
			if (book.getTitle().toLowerCase().contains(query.toLowerCase())) {
                results.add(book);
            }
        }
        return results;
    }

    public List<Book> searchByAuthor(List<Book> books, String query) {
        List<Book> results = new ArrayList<>();

        for (Book book : books) {
            if (book.getAuthor().toLowerCase().contains(query.toLowerCase())) {
                results.add(book);
            }
        }
        return results;
    }
}